### PR TITLE
Provide context on E0308 involving fn items

### LINKED ISSUE #73487: Explain why fn pointers for different fn items are different types
[Given](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=b172ce6216f978214416c1e42119bf7d)

```rust
fn foo() {}
fn bar() {}

fn main() {
    let mut x = foo;
    x = bar;
}
```

The compiler should explain why `foo` and `bar` are not the same type and should [suggest first changing `x`'s type to `fn() -> ()`](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=ecbd057607074e2a9b45308a70bebb8c).

_taken from https://twitter.com/ekuber/status/1273753857542111237_

### PR DESCRIPTION
**Pull request by estebank**
Fix #73487.

### CODE REVIEW

File: src/librustc_typeck/check/mod.rs [hunk 1]
```diff
@@ -5496,6 +5496,48 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
...
+                let sig1 = self.tcx.fn_sig(*did1).subst(self.tcx, substs1);
+                let sig2 = self.tcx.fn_sig(*did2).subst(self.tcx, substs2);
+                if sig1 != sig2 {
+                    return;
+                }
+                err.note(
+                    "different `fn` items always have unique types, even if their signatures are \
+                     the same",
+                );
+                err.help(&format!("change the expectation to require function pointer `{}`", sig1));
```
Comment (davidtwco) [inline src/librustc_typeck/check/mod.rs:5516]: Do you think it is worth de-duplicating the emission of these help messages? e.g. moving them after the `match` and returning what you need from the match arms?

File: src/test/ui/fn/fn-item-type.stderr [hunk 2]
```diff
@@ -6,34 +6,57 @@ LL |     eq(foo::<u8>, bar::<u8>);
    |
    = note: expected fn item `fn(_) -> _ {foo::<u8>}`
               found fn item `fn(_) -> _ {bar::<u8>}`
+   = note: different `fn` items always have unique types, even if their signatures are the same
+   = help: change the expectation to require function pointer `fn(isize) -> isize`
```
Comment (davidtwco) [inline src/test/ui/fn/fn-item-type.stderr:10]: I find the "expectation" terminology to be confusing here - it took me a little bit to understand what that would mean in terms of concrete changes to the source. I don't have any better suggestions though.

**Review by davidtwco — APPROVED** [review]
Looks good, a couple minor comments - if you don't think they're worth addressing then r=me.

**Review by davidtwco — APPROVED** [review]

File: src/test/ui/fn/fn-item-type.stderr (hunk 2 above)
Comment (estebank) [reply src/test/ui/fn/fn-item-type.stderr]: Changed the wording to 

```
   = help: change the expected type to be function pointer `fn(isize) -> isize`
   = help: if the expected type is due to type inference, cast the expected `fn` to a function pointer: `foo::<u8> as fn(isize) -> isize`
```

**Pull request merged**